### PR TITLE
refresh transaction from URL if needed, fix crash on key save if account is null

### LINKED ISSUE #2505: web3Middleware crashes on key.saved event if account is not set
In a race condition, if the `'key.saved'` event is emitted by the read-only web3Service before account has been set, the app crashes, because `'key.saved'` handler assumes the account is set.

### LINKED ISSUE #2506: web3Middleware does not properly retrieve transaction from URL
Currently, the `web3Middleware` only retrieves the transaction on route change, but since both `SET_PROVIDER` and `SET_NETWORK` actions will nuke transaction state, we need to retrieve it then as well. In addition, calling `web3Service.getTransaction` will trigger a race condition where transaction updates are sent prior to the transaction existing in redux, so instead, `ADD_TRANSACTION` should be dispatched and the `getTransaction` call will then happen in the correct order.

### PR DESCRIPTION
**Pull request by cellog**
# Description

If account changes, network changes, provider changes, or the actual URL changes, we need to refresh the transaction state.

In addition, if a `key.save` is emitted, there is a chance that the account is not yet set (race condition), so this removes a crash.

# Issues

Fixes #2505 
Fixes #2506 
Refs #2351

# Checklist:

- [X] 1 PR, 1 purpose: my Pull Request applies to a single purpose
  - [ ] This PR only contains configuration changes (package.json, etc.)
  - [X] This PR only contains code changes (if configuration changes are required, do a separate PR first, then re-base)
- [X] My code follows the style guidelines of this project, including naming conventions
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] If my code adds or changes components, I have written corresponding stories with Storybook
- [X] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: paywall/src/middlewares/web3Middleware.js [hunk 1]
```diff
@@ -116,10 +117,31 @@ export default function web3Middleware({ getState, dispatch }) {
...
       next(action)
 
+      const {
+        account,
+        router: {
+          location: { pathname, hash },
+        },
+        locks,
+      } = getState()
+      const { lockAddress, transaction } = lockRoute(pathname + hash)
```
Comment (julien51) [inline paywall/src/middlewares/web3Middleware.js]: Is that true on all routes? What happens on `/`?

Comment (julien51) [reply paywall/src/middlewares/web3Middleware.js]: I guess it returns null for both of these?

File: paywall/src/middlewares/web3Middleware.js [hunk 2]
```diff
@@ -134,41 +156,25 @@ export default function web3Middleware({ getState, dispatch }) {
...
 
-        const {
-          router: {
-            location: { pathname },
-          },
-        } = getState()
-
-        const { lockAddress } = lockRoute(pathname)
         if (lockAddress) {
           web3Service.getKeyByLockForOwner(lockAddress, action.account.address)
```
Comment (julien51) [inline paywall/src/middlewares/web3Middleware.js]: Combine this with line 146 above?

File: paywall/src/middlewares/web3Middleware.js [hunk 3]
```diff
@@ -134,41 +156,25 @@ export default function web3Middleware({ getState, dispatch }) {
...
 
       if (action.type === ADD_LOCK || action.type == UPDATE_LOCK) {
-        const lock = getState().locks[action.address]
-        if (getState().account) {
-          web3Service.getKeyByLockForOwner(
-            lock.address,
-            getState().account.address
-          )
+        const lock = locks[action.address]
+        if (account) {
```
Comment (julien51) [inline paywall/src/middlewares/web3Middleware.js]: is account assigned here?

File: paywall/src/middlewares/web3Middleware.js [hunk 4]
```diff
@@ -134,41 +156,25 @@ export default function web3Middleware({ getState, dispatch }) {
...
+          web3Service.getKeyByLockForOwner(lock.address, account.address)
         }
       } else if (
         action.type === LOCATION_CHANGE &&
         action.payload.location &&
         action.payload.location.pathname
       ) {
         // Location was changed, get the matching lock, if we are on a paywall page
```
Comment (julien51) [inline paywall/src/middlewares/web3Middleware.js]: Please update the comment ;)

File: paywall/src/middlewares/web3Middleware.js (hunk 1 above)
Comment (cellog) [reply paywall/src/middlewares/web3Middleware.js]: yes

File: paywall/src/middlewares/web3Middleware.js (hunk 2 above)
Comment (cellog) [reply paywall/src/middlewares/web3Middleware.js]: I tried, but these are actually quite different. 146 uses the URL lock, but this line uses the lock from the action, which may be completely different. I'll add a comment

File: paywall/src/middlewares/web3Middleware.js (hunk 3 above)
Comment (cellog) [reply paywall/src/middlewares/web3Middleware.js]: yes

**Pull request merged**